### ISSUE TITLE
Frontend server (v1)

### ISSUE THREAD

**Issue by vlukashov** [vaadin/vaadin-connect#170]
When building a Vaadin (Fusion) app for production I want to get a built-in SPA web server compatible with the way the frontend code is built, so that I am able get a single runnable .jar (deployable .war) that includes the entire app, and I can use it in production. The SPA server should be able to detect the user browser and serve the correct production frontend bundle (so that I can create several optimized frontend bundles: ES5 / ES6 / ES module and deliver smaller bundles to users with modern browsers).

### In scope (v1):
- deployment target: Servlet containers (API v3.1+)
- SPA server returns an `index.html` for any request
- SPA server allows setting URL patterns that should be resolved as static resources (potentially returning 404)
- set the base URL in the returned index.html file so that relative URLs work after page reload
- client-side bundle selection

### Out of scope (v1):
- server-side bundle selection: inject the correct frontend bundle scr into the generated index.html response based on the user agent string
- integration with Vaadin Router routes config
- inject custom title and meta tags based on the request URL
- support for Vaadin Component themes
- support for dev-mode (it's ok to use another server in the dev mode)

**Comment by Legioth** [comment]
What's the reasoning behind the decision to use user agent sniffing instead of feature detection?

Feature detection can be slightly more robust in edge cases, but requires downloading and running slightly more JS in the browser.

**Comment by vlukashov** [comment]
> What's the reasoning behind the decision to use user agent sniffing instead of feature detection?
> 
> Feature detection can be slightly more robust in edge cases, but requires downloading and running slightly more JS in the browser.

It actually makes more sense at this point to go for the **module / nomodule** approach for bundle selection (instead of implementing full-fledged client-side feature detection or backend user agent sniffing). Currently the build pipeline in the Vaadin Connect starter always produces the _module_ and _nomodule_ bundles, which fulfills the basic requirement of supporting IE11 and not shipping unnecessary polyfills for newer browsers (see the vaadin/base-starter-connect repo).

It will make sense to support an arbitrary number of bundles in the future, but I'd argue this is not a hard requirement for now.

### server-side user agent sniffing
This approach requires the backend to run the bundle selection logic based on the user agent string and alter the `index.html` file before sending it to the browsers, so that the browser receives the version of `index.html` that loads the correct JS bundle for it.

Pros:
 - _potentially_ better first page load performance than with client-side feature detection because browsers may start loading the .js bundle immediately when parser discovers it in HTML, without waiting for any JS to execute

Cons:
 - _potentially_ worse first page load performance than with client-side feature detection because `index.html` cannot be cached (different content is returned to different browsers for the same URL)
 - _potentially_ less robust than actual feature detection because it requires an accurate and up-to-date mapping of user agents to bundles (or to features)
 - _potentially_ more complicated dev server setup because it has to replicate the same server-side bundle selection logic

### client-side feature detection
This approach requires the client to run the bundle selection logic based on the JS language / runtime features, and dynamically load the correct JS bundle.

Pros:
 - _potentially_ better first page load performance than with backend user agent sniffing because `index.html` can be cached (it's the same for all browsers)
 - _potentially_ more robust than user agent sniffing
 - _potentially_ less complicated dev server setup

Cons:
 - _potentially_ worse first page load performance because browsers cannot start loading the .js bundle before executing the bundle selection JS snippet on the page

### module / nomodule bundles
This approach is a variation of client-side feature detection, but does not require any JS. It means that the number of bundles is always 2 and relies on the browsers to use the correct one based on ES modules support. 

Pros:
 - _potentially_ better first page load performance than with backend user agent sniffing because `index.html` can be cached (it's the same for all browsers)
- _potentially_ better first page load performance than with client-side feature detection because browsers may start loading the .js bundle immediately when parser discovers it in HTML, without waiting for any JS to execute
 - _potentially_ more robust than user agent sniffing
 - _potentially_ less complicated dev server setup

Cons:
 - hard limit for the number of bundles and their target JS language level

I've updated the ticket description to exclude backend user agent sniffing. This allows using either the module / nomodule approach, or client-side feature detection with JS. Initially it will be module / nomodule.

**Comment by Legioth** [comment]
I assume the module / nomodule approach refers to the concept described in https://stackoverflow.com/a/45947601/2376954. In that case, it does indeed seem like the best approach at this point.
